include/memory/memory.hh

```diff
@@ -31,20 +134,168 @@ public:
 
   Word loadWord(Addr addr);
   void storeWord(Addr addr, Word word);
+  Half loadHalf(Addr addr);
+  void storeHalf(Addr addr, Half half);
+  Byte loadByte(Addr addr);
+  void storeByte(Addr addr, Byte Byte);
 
-  template <std::forward_iterator It>
-  void storeRange(Addr start, It begin, It end) {
-    std::for_each(begin, end, [&start, this](auto curWord) {
-      storeWord(start, curWord);
-      start += kXLENInBytes;
-    });
-  }
-
-  std::size_t getCurrMemSize() const;
   void printMemStats(std::ostream &ost) const;
   const MemoryStats &getMemStats() const;
+
+  template <std::forward_iterator It>
+  void storeRange(Addr start, It begin, It end);
 };
 
+template <isSimType T, PhysMemory::MemoryOp op>
+inline T *PhysMemory::getEntity(Addr addr) {
+  if (addr % sizeof(T) || ((getOffset(addr) + sizeof(T)) > (1 << kOffsetBits)))
+    throw PhysMemory::MisAlignedAddrException(
+        "Misaligned memory access is not supported!");
+  AddrSections sections(addr);
+  auto offset = sections.offset;
+  auto isInTLB = tlb.tlbLookup(addr);
+  listIt it{};
+  if (isInTLB.has_value()) {
+    it = isInTLB.value();
+  } else {
+    it = PhysMemory::pageTableLookup<op>(sections);
+    tlb.tlbUpdate(addr, it);
+  }
+  Word *word = &it->wordStorage.at(offset / sizeof(Word));
+  Byte *byte = reinterpret_cast<Byte *>(word) + (offset % sizeof(Word));
+  return reinterpret_cast<T *>(byte);
+}
+
+template <PhysMemory::MemoryOp op>
+listIt PhysMemory::pageTableLookup(const AddrSections &sect) {
+  auto pt1Search =
+      pageTableFindPage<PTHighLvl, op>(pageTable, sect.indexPt1, sect.indexPt2);
+  if (pt1Search.has_value())
+    return pt1Search.value();
+  auto &pageTableLowLvl = pageTable.at(sect.indexPt1);
+  auto pt2Search = pageTableFindPage<PTLowLvl, op>(
+      pageTableLowLvl, sect.indexPt1, sect.indexPt2);
+  if (pt2Search.has_value())
+    return pt2Search.value();
+  return pageTableLowLvl.at(sect.indexPt2);
+}
+
+template <isPTType T, PhysMemory::MemoryOp op>
+inline std::optional<listIt>
+PhysMemory::pageTableFindPage(T &pTable, uint16_t pt1, uint16_t pt2) {
+  using MemOp = PhysMemory::MemoryOp;
+  using ptIt = typename T::iterator;
+  ptIt it_PT{};
+  if constexpr (std::is_same_v<T, PTHighLvl>) {
+    it_PT = pTable.find(pt1);
+  } else {
+    it_PT = pTable.find(pt2);
+  }
+  if (it_PT == pTable.end()) {
+    if constexpr (op == MemOp::LOAD)
+      throw PhysMemory::PageFaultException(
+          "Load on unmapped region in physical mem");
+    else {
+      return getNewPage(pt1, pt2);
+    }
+  }
+  return std::nullopt;
+}
```

Comment:
```suggestion
inline std::optional<typename T::iterator>
PhysMemory::pageTableFindPage(T &pTable, uint16_t pt) {
  using MemOp = PhysMemory::MemoryOp;
  using ptIt = typename T::iterator;
  using Mtype = typename T::mapped_type;
  auto it_PT = pTable.find(pt);
  if (it_PT != pTable.end())
    return std::nullopt;

  if constexpr (op == MemOp::LOAD)
    throw PhysMemory::PageFaultException(
          "Load on unmapped region in physical mem");
  else
    return pTable.emplace(pt, Mtype{}).first;
}
```

Reply:
This function can be static
In this case you simplify this function
And then you can write two functions which use this one
(Pseudocode)
```c++
auto &&[pt1, plvl2] = thisFunc<op>(plvl1, pt1);
auto &&[pt2, page] = thisFunc<op>(plvl2, pt2);
```